helix-core/src/normalized_url.rs

```diff
@@ -0,0 +1,175 @@
+///
+/// # Example case
+///
+/// Helix gets paths from other sources, like LSPs. In such a case, Helix can't know in advance for
+/// every source what the case will be for the prefix component on Windows (`C:` or `c:`). Since
+/// this can hinder comparisons, especially in [`Url`]s, this structure gives access to the opposite
+/// casing of the `Url` on Windows and uses it to compare.
+#[derive(Debug, Clone)]
+pub struct NormalizedUrl {
+    base: Url,
+    /// This field is only available on Windows since differences in casing on other OSes are not
+    /// limited to the prefix and should be dealt with differently.
+    #[cfg(windows)]
+    opposite_case: Option<Url>,
+}
+
+impl NormalizedUrl {
+    pub fn from_file_path(path: &Path) -> Option<Self> {
+        let path: Cow<'_, Path> = if path.is_absolute() {
+            Cow::Borrowed(path)
+        } else {
+            Cow::Owned(crate::path::get_canonicalized_path(path).ok()?)
+        };
+        // If we managed to canonicalize, this should never fail else its a bug
+        let base = Url::from_file_path(&path).unwrap();
+        #[cfg(windows)]
+        let opposite_case =
+            path_prefix_opposite_casing(&path).map(|x| Url::from_file_path(&x).unwrap());
+
+        Some(Self {
+            base,
+            #[cfg(windows)]
+            opposite_case,
+        })
+    }
+
+    /// Access the `Url` stored as the base of the normalized one
+    pub fn base(&self) -> &Url {
+        &self.base
+    }
+
+    /// See [`Url::path()`]
+    pub fn path(&self) -> &str {
+        self.base.path()
+    }
+
+    /// See [`Url::to_file_path()`]
+    #[allow(clippy::result_unit_err)]
+    pub fn to_file_path(&self) -> Result<PathBuf, ()> {
+        self.base.to_file_path()
+    }
+
+    /// See [`Url::path_segments()`]
+    pub fn path_segments(&self) -> Option<std::str::Split<'_, char>> {
+        self.base.path_segments()
+    }
+}
+
+impl PartialEq<Self> for NormalizedUrl {
+    fn eq(&self, other: &Self) -> bool {
+        // Since the other fields are derived from this one deterministically, we can simply
+        // compare `.base` to know if two `NormalizedUrl`s are equal.
+        self.base.eq(&other.base)
+    }
+}
+
+impl Eq for NormalizedUrl {}
+
+impl PartialOrd<Self> for NormalizedUrl {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        self.base.partial_cmp(&other.base)
+    }
+}
+
+impl Ord for NormalizedUrl {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.base.cmp(&other.base)
+    }
+}
+
+impl fmt::Display for NormalizedUrl {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.base.fmt(f)
+    }
+}
+
+// Interactions with the original Url type are below
+
+impl From<Url> for NormalizedUrl {
+    fn from(u: Url) -> Self {
+        let path = Path::new(u.path());
+        Self::from_file_path(path).expect("URL failed to convert to absolute path")
+    }
```

Comment:
This will panic on non-file scheme URLs, is that intended?

Reply:
Fixed